basesandbox/src/ipc/semaphore.rs

```diff
@@ -0,0 +1,125 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use nix::errno::Errno;
+use nix::libc;
+
+#[derive(PartialEq, Clone, Copy)]
+pub enum Mode {
+    CREATE,
+    OPEN,
+}
+
+pub struct SemaphoreLinker {
+    address_server: String,
+    address_client: String,
+}
+
+impl TwoWayInitialize for SemaphoreLinker {
+    type Server = Semaphore;
+    type Client = Semaphore;
+
+    fn new(_name: String) -> Self {
+        // Semaphore path stands for a virtual path, so must start with '/'
+        let address_server = format!("/{}", generate_random_name());
+        let address_client = format!("/{}", generate_random_name());
+        SemaphoreLinker {
+            address_server,
+            address_client,
+        }
+    }
+
+    fn create(&self) -> (Vec<u8>, Vec<u8>) {
+        (
+            serde_cbor::to_vec(&(&self.address_server, &self.address_client)).unwrap(),
+            serde_cbor::to_vec(&(&self.address_client, &self.address_server)).unwrap(),
+        )
+    }
+}
+
+/// Inter-process semaphore using POSIX
+pub struct Semaphore {
+    raw: *mut libc::sem_t,
+    mode: Mode,
+    name: String,
+}
+
+fn create(name: &str, mode: Mode) -> *mut libc::sem_t {
+    unsafe {
+        if Mode::CREATE == mode {
+            libc::sem_unlink(name.as_ptr() as *const i8);
+        }
+
+        let semaphore = match mode {
+            Mode::CREATE => libc::sem_open(
+                name.as_ptr() as *const i8,
+                nix::fcntl::OFlag::O_CREAT.bits(),
+                (libc::S_IRWXU | libc::S_IRWXG | libc::S_IRWXO) as libc::c_uint,
+                0,
+            ),
+            Mode::OPEN => libc::sem_open(name.as_ptr() as *const i8, 0),
+        };
+        assert_ne!(semaphore, libc::SEM_FAILED, "Failed to create semaphore: {}", Errno::last());
+        semaphore
+    }
+}
+
+impl InterProcessUnit for Semaphore {
+    fn new(data: Vec<u8>) -> Self {
+        let (address, mode): (String, u8) = serde_cbor::from_slice(&data).unwrap();
+        let mode = if mode == 1 {
+            Mode::CREATE
+        } else if mode == 2 {
+            Mode::OPEN
+        } else {
+            panic!("Invalid Semaphore argument")
+        };
+
+        Semaphore {
+            raw: create(&address, mode),
+            mode,
+            name: address,
+        }
+    }
+
+    fn ready(&mut self) {}
+}
+
+impl Semaphore {
```

Comment:
To me, it's usually not good to retain codes without actual use elsewhere. If this is a generic library, then it's ok. Otherwise, I think it's better to remove it now and readd to when strictly required.